smart-contract/program/src/processor/crank.rs

```diff
@@ -61,60 +62,90 @@ pub fn process_crank(
 ) -> ProgramResult {
     let accounts = Accounts::parse(accounts, program_id)?;
 
-    let present_time = Clock::get()?.unix_timestamp;
-
     let mut stake_pool = StakePool::get_checked(accounts.stake_pool, vec![Tag::StakePool])?;
-    let central_state = CentralState::from_account_info(accounts.central_state)?;
+    let mut central_state = CentralState::from_account_info(accounts.central_state)?;
+
+    let current_time = Clock::get()?.unix_timestamp;
+    let current_offset = (current_time - central_state.creation_time) / SECONDS_IN_DAY as i64;
+    // check if we need to do a system wide snapshot
+    if central_state.last_snapshot_offset < current_offset {
+        central_state.total_staked_snapshot = central_state.total_staked;
+        central_state.last_snapshot_offset = current_offset;
+        central_state.save(&mut accounts.central_state.data.borrow_mut())?;
+        // reset the delta for the pool that we are currently cranking. We don't want the history to influence our result
+        stake_pool.header.last_delta_update_offset = current_offset;
+        stake_pool.header.total_staked_delta = 0;
```

Comment:
I think the two lines `L76` and `L77` should be a in a different `if` statement

Reply:
Something like:

```rust
    if stake_pool.header.last_delta_update_offset < current_offset {
        // reset the delta for the pool that we are currently cranking. We don't want the history to influence our result
        stake_pool.header.last_delta_update_offset = current_offset;
        stake_pool.header.total_staked_delta = 0;
    }
```

Reply:
It seems right to me. Could you elaborate on that?

Reply:
If you have 2 pools `P1` and `P2` with `last_snapshot_offset = 0` and `current_offset = 1` but `last_snapshot_offset = 0`

- `P1` gets cranked:
    - `last_snapshot_offset` gets updated to 1
    - `last_delta_update_offset` and `total_staked_delta` get updated
- `P2` gets cranked:
    - `last_snapshot_offset = current_offset`
    - `last_delta_update_offset` and `total_staked_delta` do not get updated
 
Putting this in a different if statement like this would harmonize the crank behavior between pools

```rust
    if stake_pool.header.last_delta_update_offset < current_offset {
        stake_pool.header.last_delta_update_offset = current_offset;
        stake_pool.header.total_staked_delta = 0;
    }
```

Reply:
This unfortunatelly won't work. Imagine the following situation:

- `P1` gets cranked as in your example
- someone stakes into `P2` thereby changing the `total_staked` value recorded for this pool and the value of `total_staked_delta` as well. 
- `P2` gets cranked 
  - We are able to calculate the `total_staked` amount that was in the `P2` at the time of system-wide crank (~`P1` crank in this case)
  - If we set the `total_staked_delta=0` as you're suggesting we wouldn't be able to use it later to derive the value of `total_staked` amount that was in the `P2` at the time of system-wide crank.

Reply:
Isn't it what's already happening with `update_delta` hence why you need to make sure pool are cranked before `stake` and `unstake`?

Reply:
Nvm I think you are right, for readability and maintainability might be easier to harmonize how these fields are updated for all pools in the `crank` instruction and then make sure that all pools are cranked before `stake` and `unstake`.

The result is equivalent to what you currently have but it's just easier to read

Reply:
I don't think that the result is 100% equivalent as this approach doesn't burden the users with cranking. But you are right that it could be hidden in the bindings anyways.

If we just update these fields in the crank method and block the staking for the uncranked pools we probably wouldn't need these fields at all.

Are we sure that forcing the users to crank is a good UX? Because it would happen only very seldom.

Reply:
If the bindings are properly written there is no change to the UX

Reply:
The change has been implemented. Now all the operations that change the pool balance should be blocked when the pool is not cranked.

After discussion with @martincik we have decided that we will keep the bindings simple and atomic and handle the possible need for a crank in our app directly